### PR TITLE
Dev to Master

### PR DESCRIPTION
**Pull request by mikhailtarasevich**
Implements CRUD methods for work with database.

### CODE REVIEW

File: src/main/java/com/mikahil/tarasevich/dao/CrudDao.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.Optional;
+
+public interface CrudDao<E> {
+
+    //create
+    E save(E entity);
+    void saveAll(List<E> entity);
+
+    //read
+    Optional<List<E>> findAll();
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/CrudDao.java]: make return List<E>

File: src/main/java/com/mikahil/tarasevich/dao/GroupDao.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.mikahil.tarasevich.entity.Group;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface GroupDao extends CrudDao<Group> {
+
+    //create
+    Group save(Group group);
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/GroupDao.java]: remove CRUD methods from every interfaces exclude CrudDao<E>

File: src/main/java/com/mikahil/tarasevich/dao/StudentDao.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.Optional;
+
+public interface StudentDao extends CrudDao<Student>{
+
+    //create
+    Student save(Student student);
+    void subscribeStudentOnCourses (Student student);
+
+    //read
+    Optional<List<Student>> findAll();
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/StudentDao.java]: make return List<Strudent>

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    private static final BiConsumer<PreparedStatement, Integer> INT_PARAM_SETTER = (preparedStatement, integer) -> {
+        try {
+            preparedStatement.setInt(1, integer);
+            // preparedStatement.setObject();
+        } catch (SQLException e) {
+            //logger
+        }
+    };
+    protected static final BiConsumer<PreparedStatement, String> STRING_PARAM_SETTER = (preparedStatement, str) -> {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java]: make it private

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        this.findAllQuery = findAllQuery;
+        this.deleteByIdQuery = deleteByIdQuery;
+    }
+
+    @Override
+    public Optional<E> findById(Integer id) {
+        return findByParam(id, findByIdQuery, INT_PARAM_SETTER);
+    }
+
+    protected <P> Optional<E> findByParam(P param, String findByParam, BiConsumer<PreparedStatement,P> designatedParamSetter) {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java]: make it private, use protected findByStringParam and findByIntParam

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                }
+            }
+        } catch (SQLException e) {
+            //log
+            throw new DataBaseSqlRuntimeException("", e);
+        }
+        return Optional.empty();
+    }
+
+    public Optional<List<E>> findAll() {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java]: add @Override and return List<E>

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            return collectFoundEntitiesToList(preparedStatement);
+        } catch (SQLException e) {
+            //log
+            throw new DataBaseSqlRuntimeException("", e);
+        }
+    }
+
+    @Override
+    public void update(E entity) {
+
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java]: implement here

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                     connector.getConnection().prepareStatement(deleteByIdQuery)) {
+            designatedParamSetter.accept(preparedStatement, id);
+            preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            //log
+            throw new DataBaseSqlRuntimeException("", e);
+        }
+    }
+
+    protected Optional<List<E>> collectFoundEntitiesToList(PreparedStatement preparedStatement){
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractCrudDaoImpl.java]: make private and use findManyByStringParam findManyByIntParam and should return List<E>

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractPageableCrudDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+
+public abstract class AbstractPageableCrudDaoImpl<E> extends AbstractCrudDaoImpl <E> implements CrudPageableDao<E> {
+
+    public AbstractPageableCrudDaoImpl(ConnectorDB connector, String findByIdQuery, String findAllQuery, String deleteByIdQuery) {
+        super(connector, findByIdQuery, findAllQuery, deleteByIdQuery);
+    }
+
+    @Override
+    public List<E> findAll(int page, int itemsPerPage) { return null; }
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractPageableCrudDaoImpl.java]: implement here

File: src/main/java/com/mikahil/tarasevich/dao/impl/AbstractPageableCrudDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Override
+    public long count() {
+        return 0;
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/AbstractPageableCrudDaoImpl.java]: implement here

File: src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    private static final String DELETE_COURSE_QUERY = "DELETE FROM school.courses WHERE id=?";
+    private static final String DELETE_COURSE_DEPENDENCIES_QUERY =
+            "DELETE FROM school.student_courses WHERE course_id=?";
+
+    public CourseDaoImpl(ConnectorDB connector) {
+        super(connector, FIND_BY_ID_QUERY, FIND_ALL_QUERY, DELETE_COURSE_QUERY);
+    }
+
+    @Override
+    public Course save(Course course) {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java]: move impelentation to AbstractCrudDaoImpl, same about all crud methods in all calsses

File: src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            return makeCourseWithId(course, courseId);
+        } catch (SQLException e) {
+            //log
+            throw new DataBaseSqlRuntimeException("", e);
+        }
+    }
+
+    @Override
+    public void saveAll(List<Course> courses) {
+        try (final PreparedStatement preparedStatement = connector
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java]: move impelentation to AbstractCrudDao Impl, same about all crud methods in all calsses

File: src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+            preparedStatement.executeBatch();
+        } catch (SQLException e) {
+            //log
+            throw new DataBaseSqlRuntimeException("", e);
+        }
+    }
+
+    @Override
+    public Optional<Course> findByCourseName(String courseName) {
+        return findByParam(courseName, FIND_BY_COURSE_NAME_QUERY, STRING_PARAM_SETTER);
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java]: use findByStringParam same about all same places

File: src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Override
+    public List<Course> findCoursesByStudentId(Integer id) {
+        try (final PreparedStatement preparedStatement =
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java]: use findManyByIntegerParam same about all same places

File: src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+            preparedStatement.setInt(1, id);
+            return collectFoundEntitiesToList(preparedStatement).get();
+        } catch (SQLException e) {
+            //log
+            throw new DataBaseSqlRuntimeException("", e);
+        }
+    }
+
+    @Override
+    public void update(Course course) {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/CourseDaoImpl.java]: move imp to AbstractCrudDaoIml

File: src/main/java/com/mikahil/tarasevich/dao/impl/GroupDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class GroupDaoImpl extends AbstractPageableCrudDaoImpl<Group> implements GroupDao {
+
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/GroupDaoImpl.java]: same as at CourseDaoImpl

File: src/main/java/com/mikahil/tarasevich/dao/impl/StudentDaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Optional;
+
+public class StudentDaoImpl extends AbstractPageableCrudDaoImpl<Student> implements StudentDao {
+
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikahil/tarasevich/dao/impl/StudentDaoImpl.java]: same as at CourseDaoImpl

**Comment by mikhailtarasevich** [comment]
Снимок экрана 2022-12-17 в 01 06 39

**Comment by mikhailtarasevich** [comment]
Снимок экрана 2022-12-17 в 01 07 44

**Comment by mikhailtarasevich** [comment]
Снимок экрана 2022-12-22 в 00 32 50

**Pull request merged**